### PR TITLE
DBT-1327: Payment history

### PR DESCRIPTION
**Pull request by deevannaveed**
## Context
The purpose of this PR is to add a reuseable table with pagination and use this table for displaying the user payment invoices. 

## How to Test
- if you already have a subscription please delete it from the supabase admin dashboard `subscriptions` table.
- Now from the frontend deploy preview of this PR create a new subscription,  then go to `settings/subscriptions`, you should see the subscription invoices table, clicking the receipt button should open the receipt on stripe.

** Note: ** The invoices table will have only 1 invoice on creating a new subscription, to add more invoices here is sql statement that I use to duplicate the invoices, you can run this sql query in the SQL Editor on [Supabase staging.](https://supabase.com/dashboard/project/ppxzhhufvrffpvgraeyh/sql/new)
```sql
INSERT INTO invoices( 
   user_id, subscription_id, amount, metadata, invoice_id, invoice_pdf, invoice_url, currency
)
SELECT user_id, subscription_id, amount, metadata, invoice_id, invoice_pdf, invoice_url, currency FROM invoices limit 1 offset 0;
```

### CODE REVIEW

**Comment by abelbordonado** [comment]
Use this branch when adding th translation https://github.com/vocali-genesis/admin-portal/pull/16

File: resources/table/index.tsx [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+interface Props {
+  columns: ColumnConfig<TableDataModel>[]
+  data: TableDataModel
+  isPagination?: boolean
+}
+
+const Table: React.FC<Props> = ({ columns, data, isPagination = false,  }) => {
+
+  const [currentPage, setCurrentPage] = useState(1);
+  const itemsPerPage = 5;
```
Comment (abelbordonado) [inline resources/table/index.tsx]: We can have an optional prop for this

File: resources/table/index.tsx [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  const totalRecords = data.length;
+  const totalPages = Math.ceil(totalRecords / itemsPerPage);
+
+  const paginatedData = !isPagination ? data : data.slice(
+    (currentPage - 1) * itemsPerPage,
+    currentPage * itemsPerPage
+  );
```
Comment (abelbordonado) [inline resources/table/index.tsx]: I believe this is just to start , normally pagination is handled by the backend, you can trust that the data recived is yet paginated.

Total Records shall be a prop.

File: resources/table/pagination.tsx [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  return (
+    <div className="flex justify-between items-center mt-4 text-gray-500">
+      <div>
+        Total Records: {totalRecords}, Page {currentPage} of {totalPages}
+      </div>
+      <div className="flex items-center space-x-2">
+        <button
+          onClick={() => onPageChange(Math.max(currentPage - 1, 1))}
+          disabled={currentPage === 1}
+          className="px-4 py-2 bg-gray-200 rounded disabled:opacity-50"
```
Comment (abelbordonado) [inline resources/table/pagination.tsx]: Lets grapp the button in a PaginatedButton in  a sub function in the same page.

You can left paddings outside the component and keep them in this level.

File: resources/table/table.tsx [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        {sortedData.map((item, rowIndex) => (
+          <tr key={rowIndex} className="border-t-2 border-gray-100">
+            {columns.map((column, colIndex) => (
+              <td
+                key={colIndex}
+                className="px-6 py-4 whitespace-nowrap text-sm font-medium text-gray-900"
+              >
+                {column.render
+                  ? column.render(item)
+                  : (item[column.dataIndex!] as React.ReactNode)}
```
Comment (abelbordonado) [inline resources/table/table.tsx]: Challenges, no need now, but what about nested properties? ('item.config.value') :D

File: resources/table/table.tsx [hunk 5]
```diff
@@ -0,0 +1,87 @@
+import React, { useState } from "react";
+
+interface TableProps<T> {
```
Comment (abelbordonado) [inline resources/table/table.tsx:5]: Thought we were using tailwind css components

File: resources/table/table.tsx [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+      if (a[column.dataIndex!] < b[column.dataIndex!])
+        return direction === "asc" ? -1 : 1;
+      if (a[column.dataIndex!] > b[column.dataIndex!])
+        return direction === "asc" ? 1 : -1;
+      return 0;
+    });
+
+    setSortConfig({ key: column.dataIndex, direction });
+    setSortedData(sorted);
+  };
```
Comment (abelbordonado) [inline resources/table/table.tsx]: Sorted may be handle by the API.
Maybe we need a flag, if is not handled by api we can paginate and sort in local.
If it is, then we don't do it.

File: tailwind.config.ts [hunk 7]
```diff
@@ -5,6 +5,7 @@ const config: Config = {
     "./pages/**/*.{js,ts,jsx,tsx,mdx}",
     "./components/**/*.{js,ts,jsx,tsx,mdx}",
     "./app/**/*.{js,ts,jsx,tsx,mdx}",
+    "./resources/**/*.{js,ts,jsx,tsx,mdx}",
```
Comment (abelbordonado) [inline tailwind.config.ts:8]: Cool

File: resources/table/index.tsx (hunk 2 above)
Comment (deevannaveed) [reply resources/table/index.tsx]: Yeah, I am actually gonna try to make it in a way that supports frontend pagination and also it should be able to load do lazy loading where we just load single page data at a time.

File: resources/table/table.tsx (hunk 5 above)
Comment (deevannaveed) [reply resources/table/table.tsx:5]: Actually TailwindCSS only provides us with the CSS components, the react components by Tailwind are paid.

File: resources/table/table.tsx (hunk 4 above)
Comment (deevannaveed) [reply resources/table/table.tsx]: Yeah, in the future if we need that we can actually make it to support passing a string with with dots e.g. `item.metadata.email` etc.

File: resources/table/table.tsx (hunk 6 above)
Comment (deevannaveed) [reply resources/table/table.tsx]: Yeah, in the lazyload senario we should sort via API, from here we will just emit a event on sort change.

File: resources/table/pagination.tsx (hunk 3 above)
Comment (deevannaveed) [reply resources/table/pagination.tsx]: > You can left paddings outside the component and keep them in this level.

I didn't fully understand this line what you mean here? :D

File: core/module/services.types.ts [hunk 8]
```diff
@@ -55,4 +55,5 @@ export interface AuthService {
 export interface SubscriptionService {
   getSubscriptionLink(): Promise<{ url: string | null }>
   getActiveSubscription(): Promise<Record<string, string|number>>
+  getInvoices(): Promise<[Record<string, string | number>] | []>
```
Comment (abelbordonado) [inline core/module/services.types.ts]: You mean `Array<Record<string, string | number >>`
Maybe can make an alias tyoe for this part `Record<string, string | number >` is this a reciept?

File: modules/auth/saas/settings/subscriptions.tsx [hunk 9]
```diff
@@ -6,12 +6,103 @@ import Spinner from "@/resources/containers/spinner";
 import Service from "@/core/module/service.factory";
 import { useRouter } from "next/router";
 import moment from "moment";
+import Table from "@/resources/table";
```
Comment (abelbordonado) [inline modules/auth/saas/settings/subscriptions.tsx]: I just realize, this subscriptions is not in the `saas` settings, need to be in the `subscriptions module` 
I moved the lang of the previous PR already, lets make sure the lang file is flat (for the IA be able to manage it easily)

File: modules/auth/saas/settings/subscriptions.tsx [hunk 10]
```diff
@@ -6,12 +6,103 @@ import Spinner from "@/resources/containers/spinner";
...
+      amount: "€ 150",
+    });
+  }
+);
+
+const PaymentHistory: React.FC = () => {
+  const columns: ColumnConfig<TableDataModel>[] = [
+    { title: "Invoice ID", dataIndex: "id", sorter: true },
+    {
+      title: "Date",
```
Comment (abelbordonado) [inline modules/auth/saas/settings/subscriptions.tsx]: Remember to make the titles translated, the lang now is in the same subscription file 'subscription-settings'

File: modules/auth/saas/settings/subscriptions.tsx [hunk 11]
```diff
@@ -6,12 +6,103 @@ import Spinner from "@/resources/containers/spinner";
...
+    useState<TableDataModel[]>(intialData);
+
+  useEffect(() => {
+    setPaginatedData(() => {
+      return data.slice(
+        (currentPage - 1) * itemsPerPage,
+        currentPage * itemsPerPage
+      );
+    });
+  }, [currentPage]);
```
Comment (abelbordonado) [inline modules/auth/saas/settings/subscriptions.tsx]: Can this logic go in the table component? Is this a server or a backend pagination?

File: resources/table/table.tsx (hunk 5 above)
Comment (abelbordonado) [reply resources/table/table.tsx:5]: Sad 
Can search for others

**Review by abelbordonado — COMMENTED** [review]
Some comments

File: modules/auth/saas/settings/subscriptions.tsx (hunk 9 above)
Comment (deevannaveed) [reply modules/auth/saas/settings/subscriptions.tsx]: I think you were able to make nested string work?

File: modules/auth/saas/settings/subscriptions.tsx (hunk 10 above)
Comment (deevannaveed) [reply modules/auth/saas/settings/subscriptions.tsx]: Done, that was on my list

**Pull request merged**